mongo.go

```diff
@@ -64,40 +47,47 @@ func newItem(i *mongoItem) *resource.Item {
 	}
 }
 
-// C returns the mongo collection managed by this storage handler
-// from a Copy() of the mgo session.
-func (m *Handler) c(ctx context.Context) (*mgo.Collection, error) {
-	if err := ctx.Err(); err != nil {
-		return nil, err
-	}
-	// With mgo, session.Copy() pulls a connection from the connection pool
-	s := m.session.Copy()
-	// Ensure safe mode is enabled in order to get errors
+// Handler handles resource storage in a MongoDB collection.
+type Handler func(ctx context.Context) (*mgo.Collection, error)
+
+// NewHandler creates an new mongo handler
+func NewHandler(s *mgo.Session, db, collection string) Handler {
 	s.EnsureSafe(&mgo.Safe{})
-	// Set a timeout to match the context deadline if any
-	if deadline, ok := ctx.Deadline(); ok {
-		timeout := deadline.Sub(time.Now())
-		if timeout <= 0 {
-			timeout = 0
+	return func(ctx context.Context) (*mgo.Collection, error) {
+		if err := ctx.Err(); err != nil {
+			return nil, err
+		}
+		// Ensure safe mode is enabled in order to get errors
+		s.EnsureSafe(&mgo.Safe{})
+		// Set a timeout to match the context deadline if any
+		if deadline, ok := ctx.Deadline(); ok {
```

Comment:
Why moving this here? It will force every custom implementations to copy/past this code to keep context deadline management (which is part of the "contract" with rest-layer).